src/core/CodeSnippet.tsx

```diff
@@ -0,0 +1,542 @@
+      ...prevState,
+      activeLanguage: initialActiveLanguage,
+    }));
+
+    // Call onChange if necessary, but not on the first render
+    if (onChange && initialActiveLanguage && !isFirstRender.current) {
+      onChange(initialActiveLanguage);
+    }
+
+    // Set isFirstRender to false after the first render
+    isFirstRender.current = false;
+  }, [initialActiveLanguage, onChange]);
+
+  const codeRef = useRef<HTMLDivElement>(null);
+
+  // Check if there's a snippet available for the active language - optimized with specific dependencies
+  const hasSnippetForActiveLanguage = useMemo(() => {
+    if (!state.activeLanguage) return false;
+    return childrenArray.some(
+      (child) =>
+        isValidElement(child) && child.props.lang === state.activeLanguage,
+    );
+  }, [state.activeLanguage, childrenArray]);
+
+  // Function to get the current code text content
+  const getCodeText = useCallback((): string | null => {
+    if (
+      !state.activeLanguage ||
+      !hasSnippetForActiveLanguage ||
+      !codeRef.current
+    )
+      return null;
+
+    // Get the text content from the code element
+    const codeElement = codeRef.current.querySelector("pre");
+    return codeElement ? codeElement.textContent || "" : null;
+  }, [state.activeLanguage, hasSnippetForActiveLanguage]);
+
+  // Update selected API key if apiKeys changes
+  useEffect(() => {
+    if (
+      apiKeys &&
+      apiKeys.length > 0 &&
+      !apiKeys.includes(state.selectedApiKey)
+    ) {
+      setState((prev) => ({
+        ...prev,
+        selectedApiKey: apiKeys[0],
+      }));
+    }
+  }, [apiKeys, state.selectedApiKey]);
+
+  // Event handlers - memoized with empty dependencies as they only use setState
+  const handleSDKTypeChange = useCallback(
+    (type: SDKType) => {
+      // pick the first language that matches the new SDK prefix
+      const nextLang = languages.find((l) => l.startsWith(`${type}_`)) ?? null;
+
+      setState((prev) => ({
+        ...prev,
+        activeSDKType: type,
+        activeLanguage: nextLang,
+      }));
+      if (nextLang && onChange) onChange(nextLang);
+    },
+    [languages, onChange],
+  );
+
+  const handleLanguageChange = useCallback(
+    (language: string) => {
+      setState((prev) => ({
+        ...prev,
+        activeLanguage: language,
+      }));
+
+      // Always call onChange for explicit user changes
+      if (onChange) {
+        onChange(language);
+      }
+    },
+    [onChange],
+  );
+
+  const handleMouseEnter = useCallback(() => {
+    setState((prev) => ({
+      ...prev,
+      ui: { ...prev.ui, isHovering: true },
+    }));
+  }, []);
+
+  const handleMouseLeave = useCallback(() => {
+    setState((prev) => ({
+      ...prev,
+      ui: { ...prev.ui, isHovering: false },
+    }));
+  }, []);
+
+  const handleApiKeyChange = useCallback((apiKey: string) => {
+    setState((prev) => ({
+      ...prev,
+      selectedApiKey: apiKey,
+    }));
+  }, []);
```

Comment:
One thing I thought about a bit after the review is whether we, as in ably-ui, should be responsible for updating the code snippets (well, `children` effectively) when the selected API key changes. The current [`Code`](https://github.com/ably/docs/blob/809e93240bd14c9550cc60024ce8ae918d88cab5/src/components/blocks/software/Code/index.tsx#L68-L84) component has this kinda built in in not the nicest of ways.

The easiest way to see this is to browse docs while logged in and then you'll see your own API keys in the list and as you switch between them the code samples are _rewritten_ to have your keys.

Reply:
Yes, the api keys can be piped into the component as a prop and displayed in the list but they don't do anything currently. Given that Mark said that he wasn't planning to do anything with this yet, this is out of scope for this PR.

Reply:
Cool, I think as long as we keep it in the back of our mind that somewhere, somebody, needs to do some rewriting